src/rendergraph/opengl/backend/basenode.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <QMatrix4x4>
+
+namespace rendergraph {
+class BaseNode;
+class Engine;
+}
+
+class rendergraph::BaseNode {
+  public:
+    BaseNode() = default;
+    virtual ~BaseNode();
+
+    void setUsePreprocess(bool value) {
+        m_usePreprocess = value;
+    }
+    bool usePreprocess() const {
+        return m_usePreprocess;
+    }
+    virtual bool isSubtreeBlocked() const {
+        return false;
+    }
+    virtual void preprocess() {
+    }
+    virtual void render() {
+    }
+    virtual void initialize() {
+    }
+    virtual void resize(int, int) {
+    }
+    void setEngine(Engine* engine) {
+        m_pEngine = engine;
+    }
+    Engine* engine() const {
+        return m_pEngine;
+    }
+
+    void appendChildNode(BaseNode* pChild);
+    void removeAllChildNodes();
+    void removeChildNode(BaseNode* pChild);
+
+    BaseNode* parent() const {
+        return m_pParent;
+    }
+    BaseNode* firstChild() const {
+        return m_pFirstChild;
+    }
+    BaseNode* lastChild() const {
+        return m_pLastChild;
+    }
+    BaseNode* nextSibling() const {
+        return m_pNextSibling;
+    }
+    BaseNode* previousSibling() const {
+        return m_pPreviousSibling;
+    }
+
+  private:
+    Engine* m_pEngine{};
+    bool m_usePreprocess{};
+
+    BaseNode* m_pParent{};
+    BaseNode* m_pFirstChild{};
+    BaseNode* m_pLastChild{};
+    BaseNode* m_pNextSibling{};
+    BaseNode* m_pPreviousSibling{};
```

Comment:
can you help me understand this datastructure a little more?

So this models an N-ary tree, where the intermediary nodes are also doubly linked? So essentially like
```asciiflow
               ┌──────┐                 
               │Parent│                 
               └──▲───┘                 
                  │                     
┌───────┐         │           ┌────────┐
│next   │      ┌──┴───┐       │previous│
│Sibling◄──────┤ this ┼───────►Sibling │
└───────┘      └┬────┬┘       └────────┘
   ┌────────────┘    └───────────┐      
┌──▼──┐                       ┌──▼──┐   
│first│                       │last │   
│Child│                       │Child│   
└─────┘                       └─────┘   
```

Is this correct? If so, what relationship here is the owning one? does `this` own `firstChild`?

Reply:
Yes, this mimicks QSGNode. 
With the usual Qt pattern, a parent owns all its children (when the parent is deleted, it deletes all children).

Reply:
I added some comments.

Reply:
Please re-review and resolve conversation.

Reply:
I'll have to wrap my head around this and revisit some time else. Thanks for the comments though.